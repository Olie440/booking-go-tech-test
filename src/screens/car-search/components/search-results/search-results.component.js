import React, { Component } from 'react';
import { connect } from 'react-redux';
import { debounce } from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

import carSearch from '../../../../redux/actions/car-search';
import './search-results.style.css';

export class SearchResults extends Component {
    state = {
        input: ''
    }

    render() {
        return (
            <div className="search-results">
                <input
                    aria-label="city, airport, station, region and district..."
                    className="search-results__input"
                    id="search-results"
                    onChange={this.onInputChange}
                    value={this.state.input}
                    placeholder="city, airport, station, region and district..." />
                { this.renderResults() }
            </div>
        );
    }

    renderResults() {
        const { state, data } = this.props.results;

        if (state === 'Loading') {
            return (
                <div className="search-results__loading-spinner">
                    <FontAwesomeIcon icon={faSpinner} size="lg" spin />
                </div>
            );
        }

        if (state !== 'Success') {
            return null;
        }

        return (
            <div className="search-results__results">
                { data.map(ResultRow) }
            </div>
        );
    }

    onInputChange = (e) => {
        this.setState({ input: e.target.value });
        this.doSearch(e.target.value);
    };

    doSearch = debounce((term) => {
        this.props.carSearch(term, 6);
    }, 500);
}

const placeTypes = {
    T: 'Station',
    A: 'Airport',
    C: 'City',
    D: 'District'
}

function ResultRow({ placeType, bookingId, name, iata, city, region, country}) {
    return (
        <div className="search-results__result" key={bookingId}>
            <span className={`search-results__result-type ${placeType}`}>
                {placeTypes[placeType] || 'Other'}
            </span>
            <div className="search-results__result-cell">
                <span className="search-results__result-name">
                    {name} {iata && `(${iata})`}
                </span>
                <span className="search-results__result-location">
                    {city || region}, {country}
                </span>
            </div>
        </div>
    );
}

export function mapStateToProps({ results }) {
    return {
        results
    };
};

export default connect(mapStateToProps, { carSearch })(SearchResults);
