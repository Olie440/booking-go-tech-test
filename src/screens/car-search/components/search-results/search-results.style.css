.search-results {
    position: relative;
}

.search-results__input {
    display: inline-block;
    border: 1px solid #b4b4b4;
    border-radius: 4px;
    padding: 8px 16px;
    font-size: 16px;
    line-height: 1.5;
    width: 100%;
}

.search-results__input:focus {
    border-color: #0f6193;
    outline: 0;
}

.search-results__input::placeholder {
    font-size: 14px;
}

.search-results__results {
    background-color: #fff;
    border-radius: 4px;
    box-shadow: 0 1px 2px rgba(0,0,0,.2);
    overflow: hidden;
    width: 100%;
}

.search-results__result {
    padding: 1rem;
    cursor: pointer;
}

.search-results__result:not(:first-child) {
    border-top: 1px solid #b4b4b4;
}

.search-results__result:hover {
    background-color: #e7f4fe;
}

.search-results__result-type {
    width: 60px;
    padding: 1px 4px;
    color: #ffffff;
    text-align: center;
    background-color: #0f6193;
    border-radius: 4px;
    font-size: 12px;
    font-weight: bold;
    line-height: 1.25rem;
    display: inline-block;
    vertical-align: top;
    margin-right: 1rem;
}

.search-results__result-type.A {
    background-color: rgb(150, 20, 18);
}

.search-results__result-type.T {
    background-color: rgb(91, 91, 91);
}

.search-results__result-type.D {
    background-color: rgb(1, 124, 68);
}

.search-results__result-cell {
    display: inline-block;
}

.search-results__result-name {
    display: block;
    line-height: 1.25rem;
    font-size: 14px;
}
.search-results__result-location {
    display: block;
    line-height: 1.25rem;
    font-size: 12px;
}

.search-results__loading-spinner {
    position: absolute;
    right: 0.5rem;
    top: 0;
    line-height: 2.75rem;
}
