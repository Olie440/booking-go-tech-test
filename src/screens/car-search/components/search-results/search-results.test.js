import React from 'react';
import { shallow } from 'enzyme';
import { SearchResults, mapStateToProps } from './search-results.component';
import mockResults from '../../../../../__mocks__/mock-results';

describe('<SearchResults />', () => {
    let component, props;

    beforeEach(() => {
        props = {
            carSearch: jest.fn(),
            results: {
                state: 'None',
                data: null
            }
        };

        component = shallow(<SearchResults {...props} />);
    });

    it('renders just the input when results.state === None', () => {
        expect(component).toMatchSnapshot();
    });

    it('renders the input and loading spinner when results.state === Loading', () => {
        component.setProps({
            results: {
                state: 'Loading',
                data: null
            }
        });

        expect(component).toMatchSnapshot();
    });

    it('renders the input and results when results.state === Success', () => {
        component.setProps({
            results: {
                state: 'Success',
                data: mockResults()
            }
        });

        expect(component).toMatchSnapshot();
    });

    it('updates the state when input is changed', () => {
        component.find('.search-results__input').simulate('change', {
            target: {
                value: 'aa'
            }
        });

        expect(component.find('.search-results__input')).toMatchSnapshot();
        expect(component.state('input')).toEqual('aa');
    });

    it('calls the carSearch action when input is changed', () => {
        component.find('.search-results__input').simulate('change', {
            target: {
                value: 'aa'
            }
        });

        component.instance().doSearch.flush();

        expect(props.carSearch).toHaveBeenCalledWith('aa', 6);
    });

    it('mapStateToProps returns the result key', () => {
        const result = mapStateToProps({
            foo: 'bar',
            results: {
                state: 'None',
                data: null
            }
        });

        expect(result).toEqual({
            results: {
                state: 'None',
                data: null
            }
        });
    });
});
