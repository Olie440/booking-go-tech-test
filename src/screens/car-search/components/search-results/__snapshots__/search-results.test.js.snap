// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResults /> renders just the input when results.state === None 1`] = `
<div
  className="search-results"
>
  <input
    aria-label="city, airport, station, region and district..."
    className="search-results__input"
    id="search-results"
    onChange={[Function]}
    placeholder="city, airport, station, region and district..."
    value=""
  />
</div>
`;

exports[`<SearchResults /> renders no results found correctly 1`] = `
<div
  className="search-results__results"
>
  <div
    className="search-results__result search-results__result--no-results"
    key="No results found"
  >
    No results found
  </div>
</div>
`;

exports[`<SearchResults /> renders the input and loading spinner when results.state === Loading 1`] = `
<div
  className="search-results"
>
  <input
    aria-label="city, airport, station, region and district..."
    className="search-results__input"
    id="search-results"
    onChange={[Function]}
    placeholder="city, airport, station, region and district..."
    value=""
  />
  <div
    className="search-results__loading-spinner"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f110",
            "M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z",
          ],
          "iconName": "spinner",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size="lg"
      spin={true}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
</div>
`;

exports[`<SearchResults /> renders the input and results when results.state === Success 1`] = `
<div
  className="search-results"
>
  <input
    aria-label="city, airport, station, region and district..."
    className="search-results__input"
    id="search-results"
    onChange={[Function]}
    placeholder="city, airport, station, region and district..."
    value=""
  />
  <div
    className="search-results__results"
  >
    <div
      className="search-results__result"
      key="airport-38566"
    >
      <span
        className="search-results__result-type A"
      >
        Airport
      </span>
      <div
        className="search-results__result-cell"
      >
        <span
          className="search-results__result-name"
        >
          Manchester Airport
           
          (MAN)
        </span>
        <span
          className="search-results__result-location"
        >
          Manchester
          , 
          United Kingdom
        </span>
      </div>
    </div>
    <div
      className="search-results__result"
      key="city-2623580"
    >
      <span
        className="search-results__result-type C"
      >
        City
      </span>
      <div
        className="search-results__result-cell"
      >
        <span
          className="search-results__result-name"
        >
          Manchester
           
        </span>
        <span
          className="search-results__result-location"
        >
          Greater Manchester
          , 
          United Kingdom
        </span>
      </div>
    </div>
    <div
      className="search-results__result"
      key="district-929"
    >
      <span
        className="search-results__result-type D"
      >
        District
      </span>
      <div
        className="search-results__result-cell"
      >
        <span
          className="search-results__result-name"
        >
          Manhattan
           
        </span>
        <span
          className="search-results__result-location"
        >
          New York State
          , 
          United States of America
        </span>
      </div>
    </div>
    <div
      className="search-results__result"
      key="airport-26411"
    >
      <span
        className="search-results__result-type A"
      >
        Airport
      </span>
      <div
        className="search-results__result-cell"
      >
        <span
          className="search-results__result-name"
        >
          Managua Airport
           
          (MGA)
        </span>
        <span
          className="search-results__result-location"
        >
          Managua
          , 
          Nicaragua
        </span>
      </div>
    </div>
    <div
      className="search-results__result"
      key="airport-79938"
    >
      <span
        className="search-results__result-type A"
      >
        Airport
      </span>
      <div
        className="search-results__result-cell"
      >
        <span
          className="search-results__result-name"
        >
          Manila International Airport
           
          (MNL)
        </span>
        <span
          className="search-results__result-location"
        >
          Luzon
          , 
          Philippines
        </span>
      </div>
    </div>
    <div
      className="search-results__result"
      key="airport-23646"
    >
      <span
        className="search-results__result-type A"
      >
        Airport
      </span>
      <div
        className="search-results__result-cell"
      >
        <span
          className="search-results__result-name"
        >
          Norman Manley International Airport
           
          (KIN)
        </span>
        <span
          className="search-results__result-location"
        >
          Kingston
          , 
          Jamaica
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchResults /> updates the state when input is changed 1`] = `
<input
  aria-label="city, airport, station, region and district..."
  className="search-results__input"
  id="search-results"
  onChange={[Function]}
  placeholder="city, airport, station, region and district..."
  value="aa"
/>
`;
